src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterKeyValuePair.cs

```diff
@@ -41,12 +45,14 @@ internal sealed class JsonKeyValuePairConverter<TKey, TValue> : JsonConverter<Ke
             string propertyName = reader.GetString()!;
             if (propertyName == KeyName)
             {
-                k = ReadProperty<TKey>(ref reader, typeToConvert, options);
+                reader.Read();
+                k = JsonSerializer.Deserialize<TKey>(ref reader, options, ref state, KeyName);
```

Comment:
The re-entry overwrites `state.Current.JsonPropertyInfo` with `jsonClassInfo.PolicyProperty` which is making the KVP converter to show a different type in the exception message in case of an error in the subsequent property.

Repro:
```cs
[Fact]
public static void ConsumeIdOnKVP()
{
    string json = @"
    {
        ""Key"": ""TestKey"",
        ""foo"": ""1"",
        ""Value"": ""TestValue""
    }";

    KeyValuePair<string, string> kvp = JsonSerializer.Deserialize<KeyValuePair<string, string>>(json);
    Console.WriteLine(kvp.Key);
}
```

Reply:
What's the diff in the exception message?

Reply:
NewConverterModel: 
```
The JSON value could not be converted to System.String. Path: $.Key | LineNumber: 3 | BytePositionInLine: 22.
```
master: 
```
The JSON value could not be converted to System.Collections.Generic.KeyValuePair`2[System.String,System.String]. Path: $ | LineNumber: 3 | BytePositionInLine: 14.
```

Reply:
The latest commit will now show the same exception message as before this PR.